File: ios/Sources/GutenbergKit/Sources/Media/MediaInfo.swift
import Foundation

public struct MediaInfo: Codable {
    public let id: Int32?
    public let url: String?
    public let type: String?
    public let title: String?
    public let caption: String?
    public let alt: String?
    public let metadata: [String: String]

    private enum CodingKeys: String, CodingKey {
        case id, url, type, title, caption, alt, metadata
    }

    public init(id: Int32? = nil, url: String?, type: String?, caption: String? = nil, title: String? = nil, alt: String? = nil, metadata: [String: String] = [:]) {
        self.id = id
        self.url = url
        self.type = type
        self.caption = caption
        self.title = title
        self.alt = alt
        self.metadata = metadata
    }
}

File: android/Gutenberg/src/main/java/org/wordpress/gutenberg/media/MediaInfo.kt
package org.wordpress.gutenberg.media

import org.json.JSONArray
import org.json.JSONObject

/**
 * Mirrors `MediaInfo` on iOS — the shape the JS editor's
 * `window.blockInserter.insertMedia(...)` receiver expects per item.
 *
 * `id` is reserved for media-library round-trips that resolve a known
 * attachment; native-inserter handoffs leave it null and the JS side
 * falls back to URL-fetch-and-transform.
 */
internal data class MediaInfo(
    val id: Int? = null,
    val url: String,
    val type: String?,
    val title: String? = null,
    val caption: String? = null,
    val alt: String? = null,
    val metadata: Map<String, String> = emptyMap(),
) {
    fun toJson(): JSONObject = JSONObject().apply {
        if (id != null) put("id", id)
        put("url", url)
        if (type != null) put("type", type)
        if (title != null) put("title", title)
        if (caption != null) put("caption", caption)
        if (alt != null) put("alt", alt)
        put("metadata", JSONObject(metadata as Map<*, *>))
    }
}

internal fun List<MediaInfo>.toJsonArray(): JSONArray =
    JSONArray().also { array -> forEach { array.put(it.toJson()) } }
